File: packages/go/emitter_test.go
package causet

import "testing"

func TestEmitterOnAndEmit(t *testing.T) {
	e := NewEmitter()
	var got any
	calls := 0
	e.On("state", func(data any) {
		calls++
		got = data
	})
	e.Emit("state", map[string]any{"x": 1})
	if calls != 1 {
		t.Fatalf("calls = %d, want 1", calls)
	}
	if m, ok := got.(map[string]any); !ok || m["x"] != 1 {
		t.Fatalf("got = %#v", got)
	}
}

func TestEmitterUnsubscribe(t *testing.T) {
	e := NewEmitter()
	calls := 0
	unsub := e.On("evt", func(any) { calls++ })
	e.Emit("evt", nil)
	unsub()
	e.Emit("evt", nil)
	if calls != 1 {
		t.Fatalf("calls = %d, want 1", calls)
	}
}

func TestEmitterWildcard(t *testing.T) {
	e := NewEmitter()
	var seenType string
	e.OnAny(func(eventType string, data any) { seenType = eventType })
	e.Emit("foo", "bar")
	if seenType != "foo" {
		t.Fatalf("seenType = %q", seenType)
	}
}

func TestEmitterDoesNotCrossFireEventTypes(t *testing.T) {
	e := NewEmitter()
	calls := 0
	e.On("a", func(any) { calls++ })
	e.Emit("b", nil)
	if calls != 0 {
		t.Fatalf("calls = %d, want 0", calls)
	}
}

func TestEmitterHandlerPanicIsRecovered(t *testing.T) {
	e := NewEmitter()
	e.On("evt", func(any) { panic("boom") })
	calls := 0
	e.On("evt", func(any) { calls++ })
	e.Emit("evt", nil)
	if calls != 1 {
		t.Fatalf("second handler should still run, calls = %d", calls)
	}
}

File: packages/java/src/test/java/com/causet/sdk/EmitterTest.java
package com.causet.sdk;

import org.junit.jupiter.api.Test;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

class EmitterTest {
    @Test
    void onAndEmitInvokesHandlerWithData() {
        Emitter emitter = new Emitter();
        AtomicReference<Object> received = new AtomicReference<>();
        emitter.on("state", received::set);
        emitter.emit("state", "payload");
        assertEquals("payload", received.get());
    }

    @Test
    void unsubscribeStopsFutureCalls() {
        Emitter emitter = new Emitter();
        AtomicInteger calls = new AtomicInteger();
        Runnable unsubscribe = emitter.on("evt", d -> calls.incrementAndGet());
        emitter.emit("evt", null);
        unsubscribe.run();
        emitter.emit("evt", null);
        assertEquals(1, calls.get());
    }

    @Test
    void doesNotCrossFireEventTypes() {
        Emitter emitter = new Emitter();
        AtomicInteger calls = new AtomicInteger();
        emitter.on("a", d -> calls.incrementAndGet());
        emitter.emit("b", null);
        assertEquals(0, calls.get());
    }

    @Test
    void handlerExceptionDoesNotStopOtherHandlers() {
        Emitter emitter = new Emitter();
        AtomicInteger calls = new AtomicInteger();
        emitter.on("evt", d -> {
            throw new RuntimeException("boom");
        });
        emitter.on("evt", d -> calls.incrementAndGet());
        emitter.emit("evt", null);
        assertEquals(1, calls.get());
    }

    @Test
    void wildcardReceivesEventTypeAndData() {
        Emitter emitter = new Emitter();
        AtomicReference<String> seenType = new AtomicReference<>();
        emitter.onAny((type, data) -> seenType.set(type));
        emitter.emit("foo", "bar");
        assertEquals("foo", seenType.get());
    }

    @Test
    void noHandlersDoesNotThrow() {
        Emitter emitter = new Emitter();
        emitter.emit("nothing-registered", "x");
        assertNull(null); // reaching here without throwing is the assertion
    }
}
